src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,341 @@
+
+        /// <summary>
+        /// The number of bytes left in the packet to be consumed.
+        /// This is not necessarily the number of bytes available in the buffer.
+        /// </summary>
+        private int _packetDataLeft { get; set; } = 0;
+
+        /// <summary>
+        /// The number of bytes in the header, according to the header.
+        /// </summary>
+        private int _packetHeaderDataLength { get; set; } = 0;
+
+        private byte _packetHeaderType { get; set; } = 0;
+
+        private byte _packetStatus { get; set; }
+
+        /// <summary>
+        /// Stored internally. It will be used for tracing.
+        /// </summary>
+        private int _spid;
+
+        #endregion
+
+        #region Constructors
+
+        public TdsReadStream(Stream underlyingStream)
+        {
+            _readBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            _underlyingStream = underlyingStream;
+        }
+
+        #endregion
+
+        #region Public Properties
+        /// <inheritdoc />
+        public override bool CanRead => _underlyingStream != null;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position 
+        {
+            get => throw new NotImplementedException(); 
+            set => throw new NotImplementedException(); 
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <inheritdoc />
+        public override async ValueTask DisposeAsync()
+        {
+            await _underlyingStream.DisposeAsync();
+            _underlyingStream = null;
+            _readBuffer = null;
+        }
+
+        /// <inheritdoc />
+        public override void Flush() => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public async ValueTask<byte> PeekByteAsync(bool isAsync, CancellationToken ct)
+        {
+            // If we have logically finished reading the packet, or if we have 
+            // reached the end of the buffer, then we need to position the buffer at the beginning of next
+            // packet start.
+            await PrepareBufferIfNeeded(isAsync, ct).ConfigureAwait(false);
+            return _readBuffer[_readIndex];
+        }
+
+        /// <inheritdoc />
+        public override int Read(Span<byte> buffer)
+        {
+            int lengthToFill = buffer.Length;
+            int totalRead = 0;
+            while (lengthToFill > 0)
+            {
+                PrepareBufferIfNeeded(isAsync: false, CancellationToken.None).ConfigureAwait(false);
+
+                int lengthToCopy = MinDataAvailableBeforeRead(lengthToFill);
+                ReadOnlySpan<byte> copyFrom = _readBuffer.AsSpan(_readIndex, lengthToCopy);
```

Comment:
Parallel comment to TdsWriteStream - I think there'll be situations where we're reading large amounts of data from the network (large varbinary blobs) and we can Read straight into the buffer rather than copying twice.

Reply:
Thanks @saurabh500, I understand and agree with your point - we want to minimise Reads. Having approximately `2 * ([requested data size] / [packet size])` Read calls will become painful if we choose a 4KB packet size and want to read a 5MB payload. This'd result in 1280 packets and 2560 Read calls; it'd probably be faster than copying 4KB of memory 1280 times, but it's still hardly ideal.

A compromise might be the best of all worlds, at the expense of 8 bytes in the buffer.
* If we guarantee that there'll always be space in the buffer for 8 bytes (i.e. when preparing the buffer at the start of a new packet, reset the current position if it's < 8 bytes away from the end of the buffer) then we simplify the parsing of the packet header significantly.
* We can fill up to the end of the buffer from the network (so we know we're always reading in large chunks), copying from the buffer to the output `Memory<byte>`/`Span<byte>`
* If we reach the end of the buffer and haven't finished writing to the output, we can Read from the underlying stream to the output Memory/Span directly
  * There might be a safeguard needed here to stop us from accidentally reading from beyond the end of the TDS packet.
  * I *think* the position in the read buffer should be trivial enough to calculate as `([offset into buffer at start of Read] + MIN([length of output Span], [buffer size] - [offset into buffer at start of Read])) MOD [buffer size]`, so we can keep that in sync.